src/ApiCore/ResourceTemplate/RelativeResourceTemplate.php

```diff
@@ -0,0 +1,296 @@
+                }
+                $keys[$segment->getKey()] = true;
+            }
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function __toString()
+    {
+        return implode("/", $this->segments);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function render(array $bindings)
+    {
+        $literalSegments = [];
+        $keySegmentTuples = self::buildKeySegmentTuples($this->segments);
+        foreach ($keySegmentTuples as list($key, $segment)) {
+            /** @var Segment $segment */
+            if ($segment->getSegmentType() === Segment::LITERAL_SEGMENT) {
+                $literalSegments[] = $segment;
+                continue;
+            }
+            if (!array_key_exists($key, $bindings)) {
+                throw new ValidationException(
+                    "Rendering error - missing required binding '$key' for segment '$segment'."
+                );
+            }
+            $boundSegment = $segment->bindTo($bindings[$key]);
+            $literalSegments[] =  $boundSegment;
+        }
+        return implode("/", $literalSegments);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function matches($path)
+    {
+        try {
+            $this->match($path);
+            return true;
+        } catch (ValidationException $ex) {
+            return false;
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function match($path)
+    {
+        // High level strategy for matching:
+        // - Break $path into pieces based on '/'
+        // - Build a list of Segments from our template, where any variable segments are
+        //   flattened into a single, non-nested list
+        // - Match pieces of the path with Segments in the flattened list
+
+        $pathPieces = explode('/', $path);
+
+        // In order to build correct bindings after we match the $path against our template, we
+        // need to (a) calculate the correct positional keys for our wildcards, and (b) maintain
+        // information about the variable identifier of any flattened segments. To do this, we
+        // build a list of [string, Segment] tuples, where the string component is the appropriate
+        // key.
+        $keySegmentTuples = self::buildKeySegmentTuples($this->segments);
+        $flattenedKeySegmentTuples = self::flattenKeySegmentTuples($keySegmentTuples);
+        assert(count($flattenedKeySegmentTuples) > 0);
+
+        // We would like to match pieces of our path 1:1 with the segments of our template. However,
+        // this is confounded by the presence of double wildcards ('**') in the template, which can
+        // match multiple segments in the path.
+        // Because there can only be one '**' present, we can determine how many segments it must
+        // match by examining the difference in count between the template segments and the
+        // path pieces.
+
+        if (count($pathPieces) < count($flattenedKeySegmentTuples)) {
```

Comment:
WDYT about computing the counts once and re-using the cached value?

Reply:
Fixed in https://github.com/googleapis/gax-php/pull/209